### PR TITLE
RangeHashedDictionary use interval tree

### LINKED ISSUE #23821: range_hashed wierd issue with performance.
```sql
version 21.5.1

drop dictionary if exists curs_dict;
drop table if exists curs;

CREATE TABLE curs
(
    `a` UInt64,
    `b` UInt64,
    `e` UInt64,
    `rate` UInt16
)
ENGINE = MergeTree
ORDER BY a

insert into curs 
   select 1, 
   toYYYYMMDDhhmmss(toDateTime('2020-01-01 00:00:00')+number*2) , 
   toYYYYMMDDhhmmss(toDateTime('2020-01-01 00:00:00')+number*2+1), 1 
from numbers(1000000);

CREATE DICTIONARY curs_dict
(
    `a` Unt64,
    `b` UInt64,
    `e` UInt64,
    `rate` UInt16
)
PRIMARY KEY a
SOURCE(CLICKHOUSE(TABLE curs DB 'default' USER 'default'))
LIFETIME(MIN 0 MAX 10)
LAYOUT(RANGE_HASHED)
RANGE(MIN b MAX e);

0 rows in set. Elapsed: 115.715 sec.


SELECT sum(dictGetUInt16('curs_dict', 'rate', toUInt64(1), toYYYYMMDDhhmmss((toDateTime('2020-01-01 00:00:00') + (number * 2)) + 1))) AS s
FROM numbers(10000);

Elapsed: 0.023 sec.


SELECT sum(dictGetUInt16('curs_dict', 'rate', toUInt64(1), toYYYYMMDDhhmmss((toDateTime('2020-01-01 00:00:00') + (number * 2)) + 1))) AS s
FROM numbers(100000)

Elapsed: 2.264 sec.   / Expected  0.023*10 ~ 0.230

SELECT sum(dictGetUInt16('curs_dict', 'rate', toUInt64(1), toYYYYMMDDhhmmss((toDateTime('2020-01-01 00:00:00') + (rand())) + 1))) AS s
FROM numbers(100000)

Hangs forever
```

### PR DESCRIPTION
**Pull request by kitaisreal**
Changelog category (leave one):
- Performance Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Improve RangeHashedDictionary performance if for key there are a lot of intervals. Fixes #23821.

### CODE REVIEW

**Review by serxa — CHANGES_REQUESTED** [review]

File: src/Common/IntervalTree.h [hunk 1]
```diff
@@ -0,0 +1,567 @@
...
+        , right(right_)
+    {}
+
+    inline bool contains(IntervalStorageType point) const
+    {
+        return left <= point && point <= right;
+    }
+};
+
+template<typename IntervalStorageType>
```
[the remark this answers is not in the document]
Comment (kitaisreal) [reply src/Common/IntervalTree.h:29]: Fixed.

File: src/Common/IntervalTree.h [hunk 2]
```diff
@@ -0,0 +1,567 @@
...
+bool operator<(const Interval<IntervalStorageType> & lhs, const Interval<IntervalStorageType> & rhs)
+{
+    return std::tie(lhs.left, lhs.right) < std::tie(rhs.left, rhs.right);
+}
+
+template<typename IntervalStorageType>
+bool operator==(const Interval<IntervalStorageType> & lhs, const Interval<IntervalStorageType> & rhs)
+{
+    return std::tie(lhs.left, lhs.right) == std::tie(lhs.left, rhs.right);
```
[the remark this answers is not in the document]
Comment (kitaisreal) [reply src/Common/IntervalTree.h:38]: Fixed.

File: src/Common/IntervalTree.h [hunk 3]
```diff
@@ -0,0 +1,567 @@
...
+struct IntervalTreeVoidValue {};
+
+template <typename Interval, typename Value>
+class IntervalTree
+{
+public:
+    using IntervalStorageType = typename Interval::IntervalStorageType;
+
+    static constexpr bool is_empty_value = std::is_same_v<Value, IntervalTreeVoidValue>;
+    static constexpr bool is_nonempty_value = !std::is_same_v<Value, IntervalTreeVoidValue>;
```
[the remark this answers is not in the document]
Comment (kitaisreal) [reply src/Common/IntervalTree.h:56]: Fixed.

File: src/Common/IntervalTree.h [hunk 4]
```diff
@@ -0,0 +1,567 @@
...
+            return &getCurrentValue();
+        }
+
+        Iterator &operator++()
+        {
+            iterateToNext();
+            return *this;
+        }
+
+        // Iterator operator++(int) {
```
[the remark this answers is not in the document]
Comment (kitaisreal) [reply src/Common/IntervalTree.h:188]: Fixed.

File: src/Common/IntervalTree.h [hunk 5]
```diff
@@ -0,0 +1,567 @@
...
+        {
+            return getCurrentValue();
+        }
+
+        const IntervalWithValue & operator*() const
+        {
+            return getCurrentValue();
+        }
+
+        const IntervalWithValue *operator->()
```
[the remark this answers is not in the document]
Comment (kitaisreal) [reply src/Common/IntervalTree.h:172]: Fixed.

File: src/Common/IntervalTree.h [hunk 6]
```diff
@@ -0,0 +1,567 @@
...
+    {
+        ++intervals_size;
+        sorted_intervals.emplace_back(interval);
+    }
+
+    template <typename TValue = Value, std::enable_if_t<!std::is_same_v<TValue, IntervalTreeVoidValue>, bool> = true, typename ...Args>
+    void emplace(Interval interval, Args && ... args)
+    {
+        ++intervals_size;
+        sorted_intervals.emplace_back(interval, Value(std::forward<Args>(args)...));
```
[the remark this answers is not in the document]
Comment (kitaisreal) [reply src/Common/IntervalTree.h:69]: Fixed.

File: src/Common/IntervalTree.h [hunk 7]
```diff
@@ -0,0 +1,683 @@
...
+  *
+  * Build:
+  *
+  * To keep tree relatively balanced we can use median of all segment points.
+  * On each step build tree node with intervals. For root node input intervals are all intervals.
+  * First split intervals in 4 groups.
+  * 1. Intervals that lie that are less than median point. Interval right is less than median point.
+  * 2. Intervals that lie that are greater than median point. Interval right is less than median point.
+  * 3. Intervals that intersect node sorted by left ascending.
+  * 4. Intervals that intersect node sorted by right ascending.
```
[the remark this answers is not in the document]
Comment (kitaisreal) [reply src/Common/IntervalTree.h:88]: Fixed.

File: src/Common/IntervalTree.h [hunk 8]
```diff
@@ -0,0 +1,683 @@
...
+        assert(!tree_is_build);
+        sorted_intervals.emplace_back(interval);
+        increaseIntervalsSize();
+    }
+
+    template <typename TValue = Value, std::enable_if_t<!std::is_same_v<TValue, IntervalTreeVoidValue>, bool> = true>
+    void insert(Interval interval, const Value & value)
+    {
+        assert(!tree_is_build);
+        sorted_intervals.emplace_back(std::piecewise_construct, interval, value);
```
[the remark this answers is not in the document]
Comment (kitaisreal) [reply src/Common/IntervalTree.h:156]: Fixed.

File: src/Common/IntervalTree.h [hunk 9]
```diff
@@ -0,0 +1,683 @@
...
+    {
+        /// Before tree is build we store all intervals size in our first node to allow tree iteration.
+        ++intervals_size;
+        nodes[0].sorted_intervals_range_size = intervals_size;
+    }
+
+    std::vector<Node> nodes;
+    std::vector<IntervalWithValue> sorted_intervals;
+    size_t intervals_size = 0;
+    bool tree_is_build = false;
```
[the remark this answers is not in the document]
Comment (kitaisreal) [reply src/Common/IntervalTree.h:632]: Fixed.

**Review by serxa — APPROVED** [review]

**Pull request merged**